File: backend/app/models/expense.py
import uuid
from datetime import datetime

from sqlalchemy import BigInteger, DateTime, ForeignKey, String, Text, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column

from app.db.database import Base


class Expense(Base):
    __tablename__ = "expenses"

    id: Mapped[uuid.UUID] = mapped_column(
        primary_key=True,
        default=uuid.uuid4,
    )

    trip_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("trips.id", ondelete="CASCADE"),
        nullable=False,
    )

    wallet_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("wallets.id"),
        nullable=False,
    )

    paid_by: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id"),
        nullable=False,
    )

    amount_paise: Mapped[int] = mapped_column(
        BigInteger,
        nullable=False,
    )

    category: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
    )

    description: Mapped[str | None] = mapped_column(
        Text,
        nullable=True,
    )

    status: Mapped[str] = mapped_column(
        String(20),
        default="CONFIRMED",
        nullable=False,
    )

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        nullable=False,
    )

    idempotency_key: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    request_hash: Mapped[str] = mapped_column(
        String(64),
        nullable=False,
    )    

    __table_args__ = (
    UniqueConstraint(
        "trip_id",
        "idempotency_key",
        name="uq_expense_idempotency",
    ),
)

File: mobile/lib/models/expense.dart
class Expense {
  final String id;
  final String tripId;
  final String walletId;
  final String paidBy;
  final int amountPaise;
  final String category;
  final String? description;
  final String splitMode;
  final String status;
  final String createdAt;
  final List<ExpenseSplit> splits;

  Expense({
    required this.id,
    required this.tripId,
    required this.walletId,
    required this.paidBy,
    required this.amountPaise,
    required this.category,
    this.description,
    this.splitMode = 'EQUAL',
    required this.status,
    required this.createdAt,
    required this.splits,
  });

  factory Expense.fromJson(Map<String, dynamic> json) {
    final splitsJson = json['splits'] as List<dynamic>? ?? [];

    return Expense(
      id: json['id'],
      tripId: json['trip_id'],
      walletId: json['wallet_id'],
      paidBy: json['paid_by'],
      amountPaise: json['amount_paise'],
      category: json['category'],
      description: json['description'],
      splitMode: json['split_mode'] ?? 'EQUAL',
      status: json['status'] ?? 'CONFIRMED',
      createdAt: json['created_at']?.toString() ?? '',
      splits: splitsJson
          .map(
            (item) => ExpenseSplit.fromJson(
              Map<String, dynamic>.from(item),
            ),
          )
          .toList(),
    );
  }
}

class ExpenseSplit {
  final String memberId;
  final int amountPaise;

  ExpenseSplit({
    required this.memberId,
    required this.amountPaise,
  });

  factory ExpenseSplit.fromJson(Map<String, dynamic> json) {
    return ExpenseSplit(
      memberId: json['member_id'],
      amountPaise: json['amount_paise'],
    );
  }
}